tapdb/burn_tree.go

```diff
@@ -0,0 +1,425 @@
+var ErrNotBurn = errors.New("not a burn proof")
+
+// InsertBurns attempts to insert a set of new burn leaves into the burn tree
+// identified by the passed asset.Specifier. If a given proof isn't a true burn
+// proof, then an error is returned. This check is performed upfront. If the
+// proof is valid, then the burn leaf is inserted into the tree, with a new
+// merkle proof returned.
+func (bt *BurnUniverseTree) InsertBurns(ctx context.Context,
+	spec asset.Specifier,
+	burnLeaves ...*universe.BurnLeaf) universe.BurnLeafResp {
+
+	if len(burnLeaves) == 0 {
+		return lfn.Err[[]*universe.AuthenticatedBurnLeaf](
+			fmt.Errorf("no burn leaves provided"),
+		)
+	}
+
+	// Derive identifier (and thereby the namespace) from the
+	// asset.Specifier.
+	id, err := burnSpecifierToIdentifier(spec)
+	if err != nil {
+		return lfn.Err[[]*universe.AuthenticatedBurnLeaf](err)
+	}
+
+	// Perform upfront validation for all proofs. Make sure that all the
+	// assets are actually burns.
+	for _, burnLeaf := range burnLeaves {
+		if !burnLeaf.BurnProof.Asset.IsBurn() {
+			return lfn.Err[[]*universe.AuthenticatedBurnLeaf](
+				fmt.Errorf("%w: proof for asset %v is not a "+
+					"burn proof, has type %v",
+					ErrNotBurn,
+					burnLeaf.BurnProof.Asset.ID(),
+					burnLeaf.BurnProof.Asset.Type),
+			)
+		}
+	}
+
+	var finalResults []*universe.AuthenticatedBurnLeaf
+
+	var writeTx BaseUniverseStoreOptions
+	txErr := bt.db.ExecTx(ctx, &writeTx, func(db BaseUniverseStore) error {
+		insertedProofs := make(
+			map[universe.LeafKey]*universe.Proof, len(burnLeaves),
+		)
+
+		for _, burnLeaf := range burnLeaves {
+			leafKey := burnLeaf.UniverseKey
+
+			// Encode the burn proof to get the raw bytes.
+			var proofBuf bytes.Buffer
+			err := burnLeaf.BurnProof.Encode(&proofBuf)
+			if err != nil {
+				return fmt.Errorf("unable to encode burn "+
+					"proof: %w", err)
+			}
+			rawProofBytes := proofBuf.Bytes()
+
+			// Construct the universe.Leaf required by
+			// universeUpsertProofLeaf.
+			burnProof := burnLeaf.BurnProof
+			leaf := &universe.Leaf{
+				GenesisWithGroup: universe.GenesisWithGroup{
+					Genesis:  burnProof.Asset.Genesis,
+					GroupKey: burnProof.Asset.GroupKey,
+				},
+				RawProof: rawProofBytes,
+				Asset:    &burnLeaf.BurnProof.Asset,
+				Amt:      burnLeaf.BurnProof.Asset.Amount,
+			}
+
+			// Call the generic upsert function. MetaReveal is nil
+			// for burns, as this isn't an issuance instance. We
+			// also skip inserting into the multi-verse tree for
+			// now.
+			uniProof, err := universeUpsertProofLeaf(
+				ctx, db, id, leafKey, leaf, nil, true,
+			)
+			if err != nil {
+				return fmt.Errorf("unable to upsert burn "+
+					"leaf for key %v: %w", leafKey, err)
+			}
+
+			insertedProofs[leafKey] = uniProof
+		}
+
+		// Now, construct the AuthenticatedBurnLeaf results.
+		for _, burnLeaf := range burnLeaves {
+			uniProof, ok := insertedProofs[burnLeaf.UniverseKey]
```

Comment:
I think we should use `burnLeaf.UniverseKey.UniverseKey()` instead, which serializes and hashes the key's content to avoid issues with how public keys can be represented differently even though they are the same point in a curve